src/core/api.ts

```diff
@@ -1,14 +1,161 @@
-// ...
 
 export type CommonEventData = {
     data: Datum;
     genomicPosition: string;
 };
+
 export type MouseHoverCallback = (data: CommonEventData) => any;
 
 export interface UserDefinedEvents {
     mouseover?: MouseHoverCallback;
 }
+
+export interface Api {
+    on: (type: EVENT_TYPE, callback: MouseHoverCallback) => void;
+    zoomTo: (viewId: string, position: string, duration?: number) => void;
+    zoomToExtent: (viewId: string, duration?: number) => void;
+    zoomToGene: (viewId: string, gene: string, duration?: number) => void;
+    getViewIds: () => string[];
+    exportPNG: (transparentBackground?: boolean) => void;
+    exportPDF: (transparentBackground?: boolean) => void;
+    getCanvas: (options?: {
+        resolution?: number;
+        transparentBackground?: boolean;
+    }) => {
+        canvas: HTMLCanvasElement;
+        canvasWidth: number;
+        canvasHeight: number;
+        resolution: number;
+    };
+}
+
+export function createApi(
+    hgRef: React.RefObject<HiGlassApi | undefined> | HiGlassApi,
+    hgSpec: HiGlassSpec | undefined,
+    theme: Required<CompleteThemeDeep>
+): Api {
+    const getHg = () => {
+        // Safely get higlass API
+        if ('api' in hgRef) return hgRef;
+        if (hgRef.current) return hgRef.current;
+        throw new Error('Higlass ref not initalized');
+    };
+    const getCanvas: Api['getCanvas'] = options => {
+        const hg = getHg();
+        const resolution = options?.resolution ?? 4;
+        const transparentBackground = options?.transparentBackground ?? false;
+
+        const renderer = hg.pixiRenderer;
+        const renderTexture = PIXI.RenderTexture.create({
+            width: renderer.width / 2,
+            height: renderer.height / 2,
+            resolution
+        });
+
+        renderer.render(hg.pixiStage, renderTexture);
+
+        const canvas = renderer.plugins.extract.canvas(renderTexture);
+
+        // Set background color for the given theme in the gosling spec
+        // Otherwise, it is transparent
+        const canvasWithBg = document.createElement('canvas') as HTMLCanvasElement;
+        canvasWithBg.width = canvas.width;
+        canvasWithBg.height = canvas.height;
+
+        const ctx = canvasWithBg.getContext('2d')!;
+        if (!transparentBackground) {
+            ctx.fillStyle = theme.root.background;
+            ctx.fillRect(0, 0, canvasWithBg.width, canvasWithBg.height);
+        }
+        ctx.drawImage(canvas, 0, 0);
+
+        return {
+            canvas: canvasWithBg,
+            resolution,
+            canvasWidth: canvas.width,
+            canvasHeight: canvas.height
+        };
+    };
+    return {
+        on: (type, callback) => {
+            switch (type) {
+                case 'mouseover':
+                    return PubSub.subscribe(type, callback);
+                default: {
+                    console.warn('Subsriber not recognized.');
+                }
+            }
+        },
```

Comment:
Rather than using another ref, we can have the end users of the API handle "unsubscribing" from the global PubSub event bus.

Reply:
That's an interesting idea. Currently end-users could interact with this like:

```javascript
const api = await embed(/* ... */);
const token = api.on('mousover', console.log) // log the event payload 

// do something ...

PubSub.unsubscribe(token);
```

I would be comfortable calling this global, but it might be a bit confusing to call something outside of the API. I don't know if I've seen an `on`/`off` api, which could also be confusing, but that would look like:

```javascript
const api = await embed(/* ... */);
const token = api.on('mousover', console.log) // log the event payload 

// do something ...

api.off(token); // just wraps PubSub.unsubscribe
```

Personally, I think the most idiomatic approach would be to use `subscribe`/`unsubscribe` on the API since this pattern is common:

```javascript
const api = await embed(/* ... */);
const token = api.subscribe('mousover', console.log) // log the event payload 

// do something ...

api.unsubscribe(token); // just wraps PubSub.unsubscribe
```

Reply:
Thanks for the examples! I think I prefer the latter two slightly more than the first one as subscribe/unsubscribe can be done within the API. I saw `on`/`off` in HiGlass (https://docs.higlass.io/javascript_api.html#public.off), but since we are basically using the PubSub's `subscribe`/`unsubscribe`, I agree that the last one seems to be the most idiomatic approach.

Reply:
Yes, I think higlass `on`/`off` hooks into more than just `pub-sub-es`. In our case we are just wrapping `PubSubJS`, which lets you unsubscribe in many ways https://github.com/mroderick/PubSubJS: e.g. `token`, `function` or `topic`. 

```javascript
const api = await embed(/* .. */);

const cb = (data) => console.log(data);
const cb2 = (data) => console.log({ data });

const token = api.subscribe('mouseover', cb
const token2 = api.subscribe('mouseover', cb2);
const token3 = api.subscribe('mouseover', cb);

PubSub.unsubscribe(token); // unsubscribe specific listener
PubSub.unsubscribe(cb); // unsubscribe line 1 & 3 ( same cb )
PubSub.unsubscribe('mouseover'); // unsubscribe all
```

By aliasing `PubSub.unsubscribe` in our API, we have the freedom to change implementations in the future as well as allow/restrict what patterns of unsubscribe are allowed.

Reply:
I'm going to leave this as is (the clunkly `on` + `PubSub.unsubscribe`), and we can open a follow-up PR for deciding on the new naming API. Does that sound ok?

Reply:
Yes, that sounds good. This is something that we can revisit when we want to include more diverse APIs.

Reply:
And I like the current flexibility as I see the last example you provided.